File: java/polyglot-utils-java/src/main/java/dev/jdwilliams/utils/strings/StringUtils.java
package dev.jdwilliams.utils.strings;


public final class StringUtils {

    private StringUtils() {
        // Prevent instantiation
    }

    /**
     * Returns true if the string is null, empty, or contains only whitespace.
     */
    public static boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    /**
     * Normalizes whitespace by trimming and collapsing consecutive
     * whitespace characters into a single space.
     */
    public static String normalizeWhitespace(String value) {
        if (value == null) {
            return null;
        }
        return value.trim().replaceAll("\\s+", " ");
    }

    /**
     * Truncates a string to the specified max length.
     * If truncated, appends the supplied suffix.
     */
    public static String truncate(String value, int maxLength, String suffix) {
        if (value == null) {
            return null;
        }
        if (maxLength < 0) {
            throw new IllegalArgumentException("maxLength must be non-negative");
        }
        if (value.length() <= maxLength) {
            return value;
        }
        return value.substring(0, maxLength) + (suffix == null ? "" : suffix);
    }

    /**
     * Safely compares two strings for equality.
     */
    public static boolean equals(String a, String b) {
        if (a == null && b == null) {
            return true;
        }
        if (a == null || b == null) {
            return false;
        }
        return a.equals(b);
    }

    /**
     * Capitalizes the first character of the string.
     */
    public static String capitalize(String value) {
        if (isBlank(value)) {
            return value;
        }
        return Character.toUpperCase(value.charAt(0)) + value.substring(1);
    }


}

File: python/src/jdwilliams_utils/strings/string_utils.py
def capitalize(value: str | None) -> str | None:
    """
    Capitalizes the first character of the string and lowercases the rest.

    Args:
        value (str | None): The string to capitalize.
    """
    if value is None or value == "":
        return value
    return value[0].upper() + value[1:].lower()

def equals_nullable(a: str | None, b: str | None) -> bool:
    """
    Compares two strings for equality, handling None values.

    Args:
        a (str | None): The first string.
        b (str | None): The second string.
    """
    return (a or "") == (b or "")

def is_blank(value: str | None) -> bool:
    """
    Returns True if the string is None, empty, or contains only whitespace.

    Args:
        value (str | None): The string to check.
    """
    return value is None or value.strip() == ""

def normalize_whitespace(value: str | None) -> str | None:
    """
    Normalizes whitespace in the string by stripping leading/trailing whitespace
    and replacing multiple consecutive whitespace characters with a single space.

    Args:
        value (str | None): The string to normalize.
    """
    if value is None:
        return None
    return ' '.join(value.split())

def truncate(value: str | None, max_length: int, suffix: str | None = "") -> str | None:
    """
    Truncates the string to a maximum length, appending a suffix if truncation occurs.

    Args:
        value (str | None): The string to truncate.
        max_length (int): The maximum allowed length of the string.
        suffix (str | None): The suffix to append if truncation occurs.
    """
    if max_length < 0:
        raise ValueError("max_length must be non-negative")
    if value is None or len(value) <= max_length:
        return value
    return value[:max_length - len(suffix)] + suffix
